### PR TITLE
Configurable LED color + cleanup

### LINKED ISSUE #2296: [BUG] TFT Does not forward ACK to Octoprint
I see this was already posted as an issue and then closed with no resolution.

I have a TFT 35 3.0. TFT35_V3.0.27.x Dec 14 2021 with updated config.ini.

I have UART 3 turned on at 250K baud connected to a Raspberry Pi. If I use picoterm I can send data and it works like you think (e.g., G28 homes, M115 reports, etc.).

However, Octoprint won't connect -- sort of. It actually connects and you can see all the traffic. But it doesn't get acknowledgments so it "hangs". If you open the advanced section of the terminal and press Fake Acknowledge repeatedly it will dutifully send the commands you've queued up. So the comms are working but you don't get a response back so it waits forever unless you fake it.

The only difference I can see is that Octo puts line numbers and checksums on each line whereas G28 in picoterm is just a plain command.

I hear of people using this combination so there must be either something I have set up wrong or a regression?

### LINKED ISSUE #2377: [BUG] nozze heat button vs fan speed button
### Description

Inconvenient behaviour of the nozzle heat button on the status screen (Unified menue theme).
The current implementation is time consuming. To modify any of the nozzles' temperature I am either 
  - forced to wait a few seconds for the correct nozzle being displayed on the button or
  - I enter the menu with a randomly preselected nozzle and am forced to toggle through all temperatures (all nozzles + bed) until the correct heater is selected.
  
### Steps to reproduce

Example by a printer having 
  - two nozzles (two hotends), and 
  - three fans (fan 0, fan 1 and controller fan):
  
Every two seconds the nozzle heat button on the status screen toggles in between both tools, T0 and T1.

1. When T1 is shown press the button.
2. The heat submenu preselects tool T1.

Or:

1. When T0 is shown press the button.
2. The heat submenu preselects tool T0.

In contrast the fan button toggles in between F0, F1, CtA, CtI but regardless what fan is currently shown on the button, the fan speed submenu preselects always the last active fan (desired behaviour).

**Expected behavior**

The nozzle heat  button behaves the same way as the fan speed button, namely if clicked the submenu shall preselect the last active tool independent of what was shown on the nozzle heat button itself.

**Actual behavior**

Pressing the nozzle heat button selects the currently shown nozzle on the button (inconvenient behaviour).
 
### Hardware Variant

BIGTREETECH_TFT70_V3.0

### TFT Firmware Version & Main Board Firmware details
TFT70_V3.0.27.x

image

### LINKED ISSUE #2381: "ok" response not sent to supplementary serial ports, in certain cases
I just moved from 0.26 to 0.27 and noted that most likely there is a bug in parseAck.c
line 1242 --> ack_port_index = PORT_1

Due to this line, when I send a gcode from a different application (e.g Pronterface, 3DFox Android app, etc.) connected via UART4 (in my case BlueTooth module), the response is truncated.
In particular, only the first response line is send back to the supplementary serial port which originated the gcode command.
As a consequence, the second response line, usually the "ok" line, is missing.

This is not good: some applications, and the user him/herself, expects that "ok", to be sure that everything went well.

Expected behaviour:
M220
   FR:100%
   ok
M221
   echo:E0 Flow: 100%
   ok
M114
   X:-3.00 Y:-8.00 Z:0.02 E:0.00 Count X:-240 Y:-640 Z:4
   ok

Current (wrong) behaviour:
M220
   FR:100%
M221
   echo:E0 Flow: 100%
M114
   X:-3.00 Y:-8.00 Z:0.02 E:0.00 Count X:-240 Y:-640 Z:4

I am sure that code line was meant to do something useful as well, however is causing this problem.
Can you pls fix it?

Maybe this could solve the problem:
if (infoHost.wait==false) ack_port_index = PORT_1;

Thanks

### PR DESCRIPTION
**Pull request by digant73**
**IMPROVEMENTS:**
* Added configurable LED color: The LED color is used by features such as Event LED and PID processes. Two new params are configurable in config.ini:
  * New param "led_color" is used to set all the LED components (R, G etc...).
  * New param "led_always_on" is used to force LED on at startup or after Event LED and PID processes terminate. "led_always_on" flag can be set also from UI menu (under Screen menu).
  * LED color can be changed and stored on flash (if changed) from the already existing LEDColor menu:
    * The OFF button (black color used to switch the LED off) in LEDColor menu doesn't make the change persistent on the flash. In this way the OFF button can be used simply to temporary switch off the LED (at next TFT startup, the last set color will be used again).
    * When entering in LEDColor menu, the ON button is forced (LED is always switched on to current LED color).
* LED_Colors API: LED color functions moved from LEDColor menu to LED_Colors API just to share them with Event LED and PID features and any further feature requiring them.
* Cleanup on build scripts: Cleanup on build scripts (e.g. short_out_filename.py etc...).
* Cleanup on TFT variant template: Variant template (pin_Template.h) reviewed in order to allow inheritance for all properties and property sections ordered according to the initialization sequence used in function HW_Init and the configuration sequence in config.ini. In particular:
  * Debug macros (if any, they are pre-requisites before initializing any other module):
    *  DISABLE_JTAG()
    *  DISABLE_DEBUG()
  * Mandatory modules (they are always present in any TFT variant) such as:
    * XPT2046
    * W25Qxx
    * LCD
    * Serial ports
    * SD card
  * Optional modules (TFT variant's specific) such as:
    * USB drive
    * Power supply auto shutdown
    * Filament runout
    * Buzzer
    * LCD LED
    * Knob LED
    * Rotary encoder
    * ST7920 / HD44780 emulation support
* Cleanup on TFT specific templates: Reviewed all TFT specific templates (e.g. pin_TFT35_V3_0.h) according to new TFT variant template (described before).
* Minor cleanup: Usual indentation, empty lines etc...

**BUGFIXES:**
* Fixed TFT35_V3 variant template (bug introduced with the new GD TFT variants): LED_COLOR_PIN and NEOPIXEL_PIXELS info have been moved to proper TFT variant templates supporting them (e.g. the new GD TFT variants and E3 variant). Otherwise all the TFT variants based on TFT35_V3 (e.g. TFT35_V3 (not E3)) will be compiled with support to knob LED and neopixel (not supported by those variants).
* Fixed a possible wrong filament runout detection (in AddonHardware.c) at print start in case TFT is configured with "not inverted logic": The runout conditions are now different (one for inverted and one for not inverted logic).
* Fixed a possible missing disabling of position auto report: Added function coordinateQueryTurnOff (in coordinate.h and .c) to force disabling of position auto report, if any was previously set, independently from status variables such as coordinateQueryWait and infoHost.wait. Previously, the function coordinateQuery(0) (e.g. used at print end or exiting from Status menu) could fail on disabling the auto report (M154 S0 was not sent).
* Fixed #2377.
* Fixed #2381.
* Replaced hard coded SERIAL_PORT_2 (in interfaceCmd.c) with proper physical port: Fix needed in case more than one supplementary port is available and traffic is originated by a port different than serial port 2. This fix plus the fix for #2381 fixes #2296.
* Fixed wrong relaying of spontaneous ACK messages in parseAck.c + cleanup (comments) in parseAck.c and interfaceCmd.c: Now if TFT is not waiting for an ACK message, the ACK message is properly considered a spontaneous ACK message and it will be relayed to all supplementary serial ports (if any). Previously, even an ACK message related to a gcode originated by the TFT was considered a spontaneous ACK message (so also relayed to supplementary ports).
* Minor bugfix in Terminal.c: When the maximum number of pages was reached, the page history was reset to 1.

fixes #2377
fixes #2381
fixes #2296

**PR STATE:** Ready for merge

### CODE REVIEW

**Comment by kisslorand** [comment]
Can you please put "LED mindig BE" in Hungarian language file for "LED always ON"?

**Comment by Guilouz** [comment]
Perfect, I can close my PR.

**Comment by Guilouz** [comment]
Just try your PR but knob LED color is always green during PID.

**Comment by digant73** [comment]
@Guilouz LED and knob LED are set to green at PID completion. I didn't change that code compared to past

**Comment by Guilouz** [comment]
@digant73 But now when we start pid knob led is set to green and it’s always green during pid process, after pid done knob led go back to previous state (white).

**Comment by digant73** [comment]
very strange because at PID start both LED and knob LED should be red. At PID completion or timeout they should switch to green. And I didn't change that code in my PR

**Comment by Guilouz** [comment]
Yes Neopixel LED working as expected: White LED -> Start PID -> Red LED -> End PID -> Green LED -> Exit PID Menu -> White LED

But for Knob LED: White LED (my default color for knob led) -> Start PID -> Green LED -> End PID -> Green LED -> Exit PID Menu -> White LED

**Comment by digant73** [comment]
ok found and fixed the bug. Please download the changes and verify again

**Comment by Guilouz** [comment]
It's working now. 
Thanks.

File: TFT/src/User/Menu/Heat.c [hunk 1]
```diff
@@ -2,11 +2,18 @@
 #include "includes.h"
 
 static uint8_t tool_index = NOZZLE0;
+static uint8_t last_hotend_index = NOZZLE0;
+static uint8_t last_bed_index = BED;
 static uint8_t degreeSteps_index = 1;
 
-void heatSetCurrentIndex(uint8_t index)
+void heatSetCurrentIndex(int16_t index)
```
Comment (kisslorand) [inline TFT/src/User/Menu/Heat.c:9]: Perhaps int8_t...

**Comment by bigtreetech** [comment]
Hello, Please see [here](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/master/coding_standard.md#pull-request-standards). we recommend focusing on one thing in a PR, and different PR can be created for different things.

**Comment by digant73** [comment]
@bigtreetech Hi. Ok, no problem

**Pull request merged**